### PR TITLE
C#: Improve qldoc coverage

### PR DESCRIPTION
**Pull request by calumgrant**
Improve the qldoc coverage of our public API.

### CODE REVIEW

File: csharp/ql/src/Concurrency/Concurrency.qll [hunk 1]
```diff
@@ -10,22 +11,24 @@ class WaitCall extends MethodCall {
   Expr getExpr() { result = getArgument(0) }
 }
 
+/** An expression statement containing a `Wait` call. */
 class WaitStmt extends ExprStmt {
   WaitStmt() { getExpr() instanceof WaitCall }
 
+  /** Gets the expression this wait call is waiting on. */
```
Comment (hvitved) [inline csharp/ql/src/Concurrency/Concurrency.qll:18]: expression _that_ this?

Comment (hvitved) [reply csharp/ql/src/Concurrency/Concurrency.qll:18]: Same in two more places further down.

File: csharp/ql/src/Concurrency/Concurrency.qll [hunk 2]
```diff
@@ -37,22 +40,29 @@ class SynchronizedMethodAttribute extends Attribute {
   }
 }
 
-// A method with attribute [MethodImpl(MethodImplOptions.Synchronized)]
-class SynchronizedMethod extends Method {
+/** A method with attribute `[MethodImpl(MethodImplOptions.Synchronized)]`. */
+private class SynchronizedMethod extends Method {
   SynchronizedMethod() { getAnAttribute() instanceof SynchronizedMethodAttribute }
 
+  /** Holds if the method locks `this`. */
```
Comment (hvitved) [inline csharp/ql/src/Concurrency/Concurrency.qll:47]: the method -> this method

File: csharp/ql/src/Concurrency/Concurrency.qll [hunk 3]
```diff
@@ -37,22 +40,29 @@ class SynchronizedMethodAttribute extends Attribute {
...
 
+/** A block that is locked by a `lock` statement. */
 abstract class LockedBlock extends BlockStmt {
+  /** Holds if the `lock` statement locks `this`. */
   abstract predicate isLockThis();
 
+  /** Gets the lock variable of the `lock` statement, if any. */
   abstract Variable getLockVariable();
 
+  /** Gets the locked type of the `lock` statement, if any. */
```
Comment (hvitved) [inline csharp/ql/src/Concurrency/Concurrency.qll:62]: When can a locked expression not have a type? I.e., why the `, if any`?

File: csharp/ql/src/Documentation/Documentation.qll [hunk 4]
```diff
@@ -59,48 +61,60 @@ predicate isDocumentationNeeded(Modifiable decl) {
 class ReturnsXmlComment extends XmlComment {
   ReturnsXmlComment() { getOpenTag(_) = "returns" }
 
+  /** Holds if the element has a body at offset `offset`. */
```
Comment (hvitved) [inline csharp/ql/src/Documentation/Documentation.qll:64]: the element -> this element

File: csharp/ql/src/Documentation/Documentation.qll [hunk 5]
```diff
@@ -59,48 +61,60 @@ predicate isDocumentationNeeded(Modifiable decl) {
...
   predicate hasBody(int offset) { hasBody("returns", offset) }
 
+  /** Holds if the element is an opening tag at offset `offset`. */
```
Comment (hvitved) [inline csharp/ql/src/Documentation/Documentation.qll:67]: the element -> this element

File: csharp/ql/src/Documentation/Documentation.qll [hunk 6]
```diff
@@ -59,48 +61,60 @@ predicate isDocumentationNeeded(Modifiable decl) {
...
   predicate isOpenTag(int offset) { "returns" = getOpenTag(offset) }
 
+  /** Holds if the element is an empty tag at offset `offset`. */
```
Comment (hvitved) [inline csharp/ql/src/Documentation/Documentation.qll:70]: the element -> this element

File: csharp/ql/src/Documentation/Documentation.qll [hunk 7]
```diff
@@ -59,48 +61,60 @@ predicate isDocumentationNeeded(Modifiable decl) {
...
 }
 
 /** An XML comment containing an `<exception>` tag. */
 class ExceptionXmlComment extends XmlComment {
   ExceptionXmlComment() { getOpenTag(_) = "exception" }
 
+  /** Gets a `cref` attribute at offset `offset`, if any. */
   string getCref(int offset) { result = getAttribute("exception", "cref", offset) }
 
+  /** Holds if the element has a body at offset `offset`. */
```
Comment (hvitved) [inline csharp/ql/src/Documentation/Documentation.qll:81]: the element -> this element

File: csharp/ql/src/Documentation/Documentation.qll [hunk 8]
```diff
@@ -59,48 +61,60 @@ predicate isDocumentationNeeded(Modifiable decl) {
...
 }
 
 /** An XML comment containing a `<param>` tag. */
 class ParamXmlComment extends XmlComment {
   ParamXmlComment() { getOpenTag(_) = "param" }
 
+  /** Gets the name of this parameter at offset `offset`. */
   string getName(int offset) { getAttribute("param", "name", offset) = result }
 
+  /** Holds if the element has a body at offset `offset`. */
```
Comment (hvitved) [inline csharp/ql/src/Documentation/Documentation.qll:92]: the element -> this element

File: csharp/ql/src/Documentation/Documentation.qll [hunk 9]
```diff
@@ -59,48 +61,60 @@ predicate isDocumentationNeeded(Modifiable decl) {
...
 }
 
 /** An XML comment containing a `<typeparam>` tag. */
 class TypeparamXmlComment extends XmlComment {
   TypeparamXmlComment() { getOpenTag(_) = "typeparam" }
 
+  /** Gets the `name` attribute of this element at offset `offset`. */
   string getName(int offset) { getAttribute("typeparam", "name", offset) = result }
 
+  /** Holds if the element has a body at offset `offset`. */
```
Comment (hvitved) [inline csharp/ql/src/Documentation/Documentation.qll:103]: the element -> this element

File: csharp/ql/src/Documentation/Documentation.qll [hunk 10]
```diff
@@ -59,48 +61,60 @@ predicate isDocumentationNeeded(Modifiable decl) {
...
   predicate hasBody(int offset) { hasBody("typeparam", offset) }
 }
 
 /** An XML comment containing a `<summary>` tag. */
 class SummaryXmlComment extends XmlComment {
   SummaryXmlComment() { getOpenTag(_) = "summary" }
 
+  /** Holds if the element has a body at offset `offset`. */
```
Comment (hvitved) [inline csharp/ql/src/Documentation/Documentation.qll:111]: the element -> this element

File: csharp/ql/src/Documentation/Documentation.qll [hunk 11]
```diff
@@ -59,48 +61,60 @@ predicate isDocumentationNeeded(Modifiable decl) {
...
   predicate hasBody(int offset) { hasBody("summary", offset) }
 
+  /** Holds if the element has an open tag at offset `offset`. */
```
Comment (hvitved) [inline csharp/ql/src/Documentation/Documentation.qll:114]: the element -> this element

File: csharp/ql/src/Documentation/Documentation.qll [hunk 12]
```diff
@@ -59,48 +61,60 @@ predicate isDocumentationNeeded(Modifiable decl) {
...
   predicate isOpenTag(int offset) { "summary" = getOpenTag(offset) }
 
+  /** Holds if the element is empty at offset `offset`. */
```
Comment (hvitved) [inline csharp/ql/src/Documentation/Documentation.qll:117]: the element -> this element

File: csharp/ql/src/semmle/code/cil/Instructions.qll [hunk 13]
```diff
@@ -246,111 +285,137 @@ module Opcodes {
...
   class Bge_un extends BinaryBranch, @cil_bge_un {
     override string getOpcodeName() { result = "bge.un" }
   }
 
+  /** A `bge` instruction. */
   class Bge extends BinaryBranch, @cil_bge {
     override string getOpcodeName() { result = "bge" }
   }
 
+  /** A `bne.un` instruction. */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/cil/Instructions.qll:368]: bne.un.s

File: csharp/ql/src/semmle/code/cil/Instructions.qll [hunk 14]
```diff
@@ -367,62 +432,73 @@ module Opcodes {
...
   }
 
+  /** A `cgt` instruction. */
   class Cgt extends ComparisonOperation, @cil_cgt {
     override string getOpcodeName() { result = "cgt" }
   }
 
+  /** A `clt.un` instruction. */
   class Clt_un extends ComparisonOperation, @cil_clt_un {
     override string getOpcodeName() { result = "cgt.un" }
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/cil/Instructions.qll:469]: (Unrelated) `result = "clt.un"`

File: csharp/ql/src/semmle/code/cil/Instructions.qll [hunk 15]
```diff
@@ -246,111 +285,137 @@ module Opcodes {
...
   }
 
+  /** A `bgt` instruction. */
   class Bgt extends BinaryBranch, @cil_bgt {
     override string getOpcodeName() { result = "bgt" }
   }
 
+  /** A `bgt.in.s` instruction. */
   class Bgt_in_s extends BinaryBranch, @cil_bgt_un_s {
     override string getOpcodeName() { result = "bgt.un.s" }
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/cil/Instructions.qll:410]: (Unrelated) `result = "bgt.in.s"`

File: csharp/ql/src/semmle/code/cil/Instructions.qll [hunk 16]
```diff
@@ -734,449 +846,524 @@ module Opcodes {
...
   }
 
+  /** An `ldelem` instruction. */
   class Ldelem extends ReadArrayElement, @cil_ldelem {
     override string getOpcodeName() { result = "ldelem" }
 
     override Type getType() { result = getAccess() }
   }
 
+  /** An `ldelem_ref` instruction. */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/cil/Instructions.qll:856]: ldelem.ref

File: csharp/ql/src/semmle/code/cil/Instructions.qll [hunk 17]
```diff
@@ -734,449 +846,524 @@ module Opcodes {
...
   }
 
+  /** A `conv.r8` instruction. */
   class Conv_r8 extends Conversion, @cil_conv_r8 {
     override string getOpcodeName() { result = "conv.r8" }
 
     override DoubleType getType() { exists(result) }
   }
 
+  /** A `conv.r8.un` instruction. */
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/cil/Instructions.qll:1209]: conv.r.un

File: csharp/ql/src/semmle/code/cil/Instructions.qll [hunk 18]
```diff
@@ -734,449 +846,524 @@ module Opcodes {
...
   }
 
+  /** An `stind.r4` instruction. */
   class Stind_r4 extends StoreIndirect, @cil_stind_r4 {
     override string getOpcodeName() { result = "stind.r4" }
   }
 
+  /** An `stind.r8` instruction. */
   class Stind_r8 extends StoreIndirect, @cil_stind_r8 {
     override string getOpcodeName() { result = "stind.r4" }
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/cil/Instructions.qll:1330]: (Unrelated) `result = "stind.r8"`

File: csharp/ql/src/semmle/code/cil/Instructions.qll [hunk 19]
```diff
@@ -6,85 +6,98 @@ private import CIL
 private import semmle.code.dotnet.Variable as DotNet
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/cil/Instructions.qll:6]: Could we perhaps add a link to something like https://en.wikipedia.org/wiki/List_of_CIL_instructions to the file QLDoc?

File: csharp/ql/src/Concurrency/Concurrency.qll (hunk 3 above)
Comment (calumgrant) [reply csharp/ql/src/Concurrency/Concurrency.qll:62]: Because it only gives results for `lock(typeof(Foo))`.

**Pull request merged**